pkg/controller/podautoscaler/horizontal.go

```diff
@@ -419,35 +433,67 @@ func (a *HorizontalController) reconcileAutoscaler(hpav1Shared *autoscalingv1.Ho
+			// make sure we aren't below our minimum
+			setCondition(hpa, autoscalingv2.DesiredOutsideRange, v1.ConditionTrue, "TooFewReplicas", "the desired replica count was less than the minimum replica count")
+			desiredReplicas = *hpa.Spec.MinReplicas
+		case desiredReplicas > hpa.Spec.MaxReplicas:
+			// make sure we aren't above our maximum
+			setCondition(hpa, autoscalingv2.DesiredOutsideRange, v1.ConditionTrue, "TooManyReplicas", "the desired replica count was more than the maximum replica count")
+			desiredReplicas = hpa.Spec.MaxReplicas
+		default:
+			// mark that we're within acceptible limits
+			setCondition(hpa, autoscalingv2.DesiredOutsideRange, v1.ConditionFalse, "DesiredWithinRange", "the desired replica count is within the acceptible range")
 		}
 
-		if desiredReplicas > hpa.Spec.MaxReplicas {
-			desiredReplicas = hpa.Spec.MaxReplicas
+		rescale = a.shouldScale(hpa, currentReplicas, desiredReplicas, timestamp)
+		backoffDown := false
+		backoffUp := false
+		if hpa.Status.LastScaleTime != nil {
+			if !hpa.Status.LastScaleTime.Add(a.downscaleForbiddenWindow).Before(timestamp) {
+				setCondition(hpa, autoscalingv2.InBackoff, v1.ConditionTrue, "BackoffDownscale", "the time since the previous scale is still within the downscale forbidden window")
+				backoffDown = true
+			}
+
+			if !hpa.Status.LastScaleTime.Add(a.upscaleForbiddenWindow).Before(timestamp) {
+				backoffUp = true
+				if backoffDown {
+					setCondition(hpa, autoscalingv2.InBackoff, v1.ConditionTrue, "BackoffBoth", "the time since the previous scale is still within both the downscale and upscale forbidden windows")
+				} else {
+					setCondition(hpa, autoscalingv2.InBackoff, v1.ConditionTrue, "BackoffUpscale", "the time since the previous scale is still within the upscale forbidden window")
+				}
+			}
 		}
 
-		// Do not upscale too much to prevent incorrect rapid increase of the number of master replicas caused by
-		// bogus CPU usage report from heapster/kubelet (like in issue #32304).
-		if scaleUpLimit := calculateScaleUpLimit(currentReplicas); desiredReplicas > scaleUpLimit {
-			desiredReplicas = scaleUpLimit
+		if !backoffDown && !backoffUp {
+			// mark that we're not backing off
+			setCondition(hpa, autoscalingv2.InBackoff, v1.ConditionFalse, "ReadyForNewScale", "the last scale time was sufficiently old as to warrant a new scale")
 		}
 
-		rescale = a.shouldScale(hpa, currentReplicas, desiredReplicas, timestamp)
 	}
 
 	if rescale {
 		scale.Spec.Replicas = desiredReplicas
 		_, err = a.scaleNamespacer.Scales(hpa.Namespace).Update(hpa.Spec.ScaleTargetRef.Kind, scale)
 		if err != nil {
 			a.eventRecorder.Eventf(hpa, v1.EventTypeWarning, "FailedRescale", "New size: %d; reason: %s; error: %v", desiredReplicas, rescaleReason, err.Error())
+			setCondition(hpa, autoscalingv2.CanAccessScale, v1.ConditionFalse, "FailedUpdate", "the HPA controller was unable to update the target scale: %v", err)
```

Comment:
We have "FailedRescale" event, which for me is better name than "FailedUpdated".